decoder/context.go

```diff
@@ -16,8 +22,117 @@ type DecoderContext struct {
 	// CodeLenses represents a slice of executable lenses
 	// which will be executed in the exact order they're declared
 	CodeLenses []lang.CodeLensFunc
+
+	// CompletionHooks represents a map of available hooks for completion.
+	// One can register new hooks by adding an entry to this map. Inside the
+	// attribute schema, one can refer to the hooks map key to enable the hook
+	// execution on CompletionAtPos.
+	CompletionHooks CompletionFuncMap
+	// CompletionResolveHooks represents a map of available hooks for
+	// completion candidate resolving. One can register new hooks by adding an
+	// entry to this map. On completion candidate creation, one can specify a
+	// resolve hook by using the map key string. If a completion candidate has
+	// a resolve hook, ResolveCandidate will execute the hook and update the
+	// completion item.
+	CompletionResolveHooks CompletionResolveFuncMap
+}
+
+func NewDecoderContext() DecoderContext {
+	return DecoderContext{
+		CompletionHooks:        make(CompletionFuncMap),
+		CompletionResolveHooks: make(CompletionResolveFuncMap),
+	}
 }
 
 func (d *Decoder) SetContext(ctx DecoderContext) {
 	d.ctx = ctx
 }
+
+type CompletionFunc func(ctx context.Context, value cty.Value) ([]Candidate, error)
```

Comment:
As mentioned in the other PR review
 - https://github.com/hashicorp/terraform-ls/pull/1017#discussion_r933310836
 - https://github.com/hashicorp/terraform-ls/pull/1017#discussion_r933313834

is it worth bringing some of the docs here, so we can omit them downstream?

Reply:
That's a valid point, I moved it here.

I'm just wondering how easy the docs are discoverable, one won't be able to see them anywhere on hover in the LS.

Reply:
You could add a reference or comment to `CompletionFuncMap` which is easier to discover downstream on hover.